Open Positions

mdavid217 opened the pull request:
Here is the PR for the Open Positions #32 feature request. This does incorporate the correction regarding the use of an operation inside of the RP2Decimal instantiation, and also I've converted several magic numbers from when I was working on the code to proper defines. For now I've left some of the variable names shortened due to the wrapping that resulted from extending them. At least to these eyes, that reduced readability. Cheers!

mdavid217 commented on src/rp2/plugin/report/us/open_positions.py:
```diff
@@ -120,9 +122,9 @@ def generate(
             unit_cost_basis = net_cost / tot_crypto_bal
             if unit_cost_basis >= _FIAT_UNIT_WINDOW_2_DECIMAL:
-                unit_cost_basis = RP2Decimal(round(unit_cost_basis, 2))
+                unit_cost_basis = RP2Decimal(unit_cost_basis).quantize(RP2Decimal("1.00"))
```
mypy does complain about this, regardless of if I use Decimal or RP2Decimal inside of the quantize. Seems to work, just not sure if there's a better way to express this to make mypy happy.
```
src\rp2\plugin\report\us\open_positions.py:125:35: error: Incompatible types in assignment (expression has type "Decimal", variable has type "RP2Decimal")  [assignment]
                    unit_cost_basis = RP2Decimal(unit_cost_basis).quantize(RP2Decimal("1.00"))
                                      ^
```

eprbell replied:
It should be:
```suggestion
                unit_cost_basis = unit_cost_basis.quantize(FIAT_DECIMAL_MASK)
```
You'll also need to import FIAT_DECIMAL_MASK from rp2.rp2_decimal

mdavid217 replied:
I had to tweak this a little. First, since I'm using 2 and 4 decimal output, I decided not to use the define from rp2_decimal for clarity, but that pointer was useful for taking a 2nd fresh look at the RP2Decimal class. I noticed quantize is used a test there, but as an assignment, mypy complains about assigning a decimal result to an RP2Decimal variable so the second tweak to the above was to init the statement result to a new RP2Decimal object. Works correctly and satisfies mypy!

mdavid217 commented on src/rp2/plugin/report/us/open_positions.py:
```diff
@@ -120,9 +122,9 @@ def generate(
+                unit_cost_basis = RP2Decimal(unit_cost_basis).quantize(RP2Decimal("1.00"))
             elif unit_cost_basis >= _FIAT_UNIT_WINDOW_4_DECIMAL:
-                unit_cost_basis = RP2Decimal(round(unit_cost_basis, 4))
+                unit_cost_basis = RP2Decimal(unit_cost_basis).quantize(RP2Decimal("1.0000"))
```
Same as above.

mdavid217 commented on src/rp2/plugin/report/us/open_positions.py:
```diff
@@ -0,0 +1,232 @@
+            ComputedData.type_check("computed_data", computed_data)
+
+            net_cost = RP2Decimal("0")
+            for in_xact in computed_data.in_transaction_set:
```
mypy spits this error for line 90 and also 106 below. Not sure the proper way to satisfy mypy here.
```
src\rp2\plugin\report\us\open_positions.py:90:13: error: Incompatible types in assignment (expression has type "AbstractEntry", variable has type "InTransaction")  [assignment]
                for in_xact in computed_data.in_transaction_set:
                ^
```

eprbell replied:
This is because in_transaction_set is of type TransactionSet and an iterator over it returns objects of type AbstractTransaction. You can do this:
```suggestion
            for current_transaction in computed_data.in_transaction_set:
                 in_xact = cast(InTransaction, current_transaction)
```

mdavid217 replied:
Perfect, thanks! That makes sense, useful to understand.

eprbell commented:
From a first glance, it looks pretty useful! Will do a detailed review in the weekend.

mdavid217 commented:
> From a first glance, it looks pretty useful! Will do a detailed review in the weekend.

Thanks! I had some fun writing this, and it felt great to have a much faster way to get complete and accurate information that otherwise takes a long time to manually collect. The value of being able to use the data I already need to gather for taxes is a huge boon to me!

eprbell reviewed (changes requested):
This is a cool plugin, thanks for proposing it. If you want, you may even extend it into tax loss harvesting (in a separate PR). There are a few changes that are needed but overall it looks very promising!

A few comments on the template file (since I couldn't comment inline I'll add it here):
* I think the word you need is "unrealized", not "net"
* for consistency with other plugin outputs: use full words
* for consistency with other plugin outputs: use % at the end, not pct
* "calc" is unnecessary (you already distinguish it by using yellow)

So here's the list of string changes:
* USD Net Cost Basis -> USD Unrealized Cost Basis
* Pct Portfolio CB Weigth -> Cost Basis Weight %
* USD Per Unit Calc Price -> Input Price
* USD Net Calc Value -> USD Unrealized Value
* USD Net Gain / Loss -> USD Unrealized Gain / Loss
* Pct Portfolio CB Gain / Loss -> Unrealized Gain / Loss Over Total Cost Base
* Pct Portfolio Calc Weight -> USD Unrealized Value Weight %

I'm not very sure if the "Pct Portfolio CB Gain / Loss" column is very useful: do we need it and how do you use it?

You'll need a new column for "holder": otherwise if people file together you get unintended results (e.g. all their BTC are summed up together and in the "Asset - Exchange" sheet you may see two Coinbase lines, which is confusing).

Finally, for consistency with other plugins:
* "AssetPrice" -> "Asset Price"
* "Asset_and_Exchange" -> "Asset - Exchange"

Since the "Asset" sheet is really a place where the user inputs data, it might be clearer to call it "Input".

mdavid217 commented on src/rp2/plugin/report/us/open_positions.py:
```diff
@@ -0,0 +1,233 @@
+                raise RP2TypeError(f"Parameter 'asset' has non-string value {asset}")
+            ComputedData.type_check("computed_data", computed_data)
+
+            net_cost = RP2Decimal("0")
```
switched to total_cost_basis and asset_cost_basis

mdavid217 commented on src/rp2/plugin/report/us/open_positions.py:
```diff
@@ -0,0 +1,233 @@
+                in_xact = cast(InTransaction, current_transaction)
+                net_cost += in_xact.fiat_in_with_fee * (RP2Decimal("1") - computed_data.get_in_lot_sold_percentage(in_xact))
+
+            if net_cost < _FIAT_EMPTY_BAL_CUTOFF:
```
The statement and others like it are now checking against ZERO rather than a cutoff value (exclude closed positions and empty wallets from the report). Originally I used cutoffs to get around #26 but I was able to munge my input data to prevent phantom crypto dust from adding an extraneous value to the output.

mdavid217 commented on src/rp2/plugin/report/us/open_positions.py:
```diff
@@ -0,0 +1,233 @@
+            ap_sheet.append_rows(1)
+            ap_row_index: int = row_indexes["AssetPrice"]
+            self._fill_cell(ap_sheet, ap_row_index, 0, asset)
+            self._fill_cell(ap_sheet, ap_row_index, 1, 0)
```
Since there's a VLOOKUP on the two report tabs, I don't want to inadvertently instruct the user to key in prices directly to the report tabs. Instead I've chosen to use "See Input Tab" to direct their attention there.

eprbell replied:
Oh, my bad. I meant: add the "enter price" string in the input tab.

mdavid217 commented on src/rp2/plugin/report/us/open_positions.py:
```diff
@@ -0,0 +1,233 @@
+        LOGGER.info("Plugin '%s' output: %s", __name__, Path(output_file.docname).resolve())
+
+
+def main() -> None:
+    pass
+
+
+if __name__ == "__main__":
+    main()
```
This block of code exists in the other 2 reports - I used the tax report as a template when I started. Should this be removed from those reports as well?

eprbell replied:
True and yes, please! :-)

mdavid217 commented on src/rp2/plugin/report/us/open_positions.py:
```diff
@@ -0,0 +1,233 @@
+
+            # Sum the total amount of crypto asset remaining.
+            tot_crypto_bal = RP2Decimal("0")
+            for balset in computed_data.balance_set:
```
Fixed the varname. Will have to spend a little extra time making this work per user. Can we make that a separate PR or would you prefer leaving it as part of this one?

eprbell replied:
I think it's worth spending a little extra time on this one, otherwise people filing together get confusing results.

mdavid217 replied:
That does make sense. I've started refactoring the plugin to account for this. 

I set up a test with two users, e.g. a second user buys, receives a transfer, and sells BTC interspersed with the first user's activity. It appears as though the final balance on the tax form accurately tracks the balance for each user, but sales (at least in a FIFO scenario) are top-down from the combined assets rather than e.g. in processing User2's sale first hunting to their buy and drawing from that purchase first. 

I just want to confirm this is expected - if so, it does make adding multiple users to my report easier - as I can keep total cost basis as the sum for all holders of an asset and then divide that out over their respective final balances. 

My original concern was that this wouldn't be purely representational if they were individuals tracking their cost basis, but if the above is accurate, then it's OK that the holders all have the same unit cost basis, just with different crypto balances / allocations.

Can you confirm the function? Hopefully what I'm asking makes sense. :)

eprbell replied:
Yes, that sounds correct: they are filing together so their per-unit cost basis is the same but they have different balances.

mdavid217 replied:
Changes pushed incorporating multiple holders and other changes previously marked resolved since last review.

mdavid217 commented on src/rp2/plugin/report/abstract_ods_generator.py:
```diff
@@ -54,7 +55,14 @@ def _initialize_output_file(
             output_file_path.unlink()
 
-        template_path: str = str(Path(os.path.dirname(__file__)).absolute() / Path("".join(["data/template_", country.country_iso_code, ".ods"])))
+        if len(template_file_prefix) > 0 and template_file_prefix.startswith("_") is False:
+            template_file_prefix = "_" + template_file_prefix
+        template_path: str = str(
+            Path(os.path.dirname(__file__)).absolute() / Path("".join(["data/template", template_file_prefix, "_", country.country_iso_code, ".ods"]))
+        )
+        if Path(template_path).exists() is False:
+            raise Exception(f"Error: template file '{template_path}' not found.")
+
```
The main reason I put in the code this way was because I wasn't sure if you wanted to leave the country code bit in the generator, or push that responsibility up to the individual reports to do or enforce. I opted for the lighter touch, but if that works for you I can do the change and put the country code bit in the other reports.

Alternatively, if you're inclined, you can update the generator and  2 reports how you'd like it to work, and I can follow suit.

eprbell replied:
Gotcha, good point: let me work on this. I'll update the code shortly.

eprbell replied:
I uploaded the new logic: now plugins pass the full path to the template file to `_initialize_output_file()`

mdavid217 replied:
Excellent, got it and applied.

mdavid217 commented:
> I'm not very sure if the "Pct Portfolio CB Gain / Loss" column is very useful: do we need it and how do you use it?

I changed the description in the legend to be more meaningful. "Indicates the percent of which this asset’s unrealized gain/loss contributed to the whole portfolio’s Unrealized Gain / Loss %. (Gain/Loss relative to total cash investment)". 

I use this field along with the other 3 percentage fields (CB Weight %, Gain / Loss %, Unrealized Weight %) especially in weighing the absolute performance or lack thereof of an asset to the other assets in the portfolio. I especially compare this figure to the unrealized weight %. I use the information to help me chose where to allocate fresh funds without getting lost in the dollars and cents. I don't know if other people do this, or if it's just me, but it is nice seeing how each asset individually affects the entire portfolio's performance.

> You'll need a new column for "holder": otherwise if people file together you get unintended results (e.g. all their BTC are summed up together and in the "Asset - Exchange" sheet you may see two Coinbase lines, which is confusing).

Good point. I didn't think to code this in since I've only done 1 holder per file so far. This might take additional refactoring. Per a comment on one of the items - do you want me to try to put that into this PR, or do a separate PR with that?

The non-resolved items above I am requesting additional feedback on. Thanks for taking the time. -MD

eprbell reviewed (changes requested):
Looking better and better! Let me know if I any of your questions fell through the cracks and you're still waiting for me on something.

Adding here comments on template file: 
* cell "of Total Cost Basis %" is in a different font size: best to leave it in the same font size as the rest. If the user wishes, they can adjust column size with: Edit / Select All + Format / Columns / Optimal Width.

Also add a small section for the plugin in docs/output_files.md: just a few lines describing what it does and directing the user to fill the input tab.

eprbell commented:
One more thought: I can see this plugin evolving into something more than open positions. For example, it could offer advice on tax loss harvesting or other investment management issues. Should we give it a more open-ended name, like investment_report?

mdavid217 commented on src/rp2/plugin/report/us/open_positions.py:
```diff
@@ -0,0 +1,274 @@
+            unit_cost_basis: RP2Decimal = asset_cost_basis / total_crypto_balance
+            unit_data_style: str = "fiat"
+            if _FIAT_UNIT_DATA_STYLE_4_DECIMAL_MINIMUM <= unit_cost_basis < _FIAT_UNIT_DATA_STYLE_2_DECIMAL_MINIMUM:
+                unit_data_style = "fiat_unit_4"
```
Subjectively chosen based on my experience with cryptos having unit prices across the spectrum.

My intent is to show enough precision that someone looking at a chart or exchange or other resource sees enough precision in their values to meaningfully assess what they're doing while also not showing excess precision that detracts from the readability of the report. For me, that's 2 decimals $1+, and 4 decimals for most of the cryptos under a dollar, except for the super cheap ones like SHIB where the unit cost is incredibly low. 7 or 8 digits there seems sufficient.

mdavid217 commented on src/rp2/plugin/report/us/open_positions.py:
```diff
@@ -0,0 +1,274 @@
+            input_sheet.append_rows(1)
+            input_row_index: int = row_indexes["Input"]
+            self._fill_cell(input_sheet, input_row_index, 0, asset)
+            self._fill_cell(input_sheet, input_row_index, 1, "See Input Tab", data_style="fiat_unit_7")
```
The Asset and Asset - Exchange tab have VLOOKUPs that present whatever is in the Input tab. I didn't use "Enter {asset} price" due to the risk a user would interpret that as literally entering the price into the Asset or Asset - Exchange tab directly, thus overwriting the VLOOKUP formula and causing them additional work they need not do. 

Initially I figured using a default of 0, null, or "See Input Tab" would be less misleading, but I looked, and I could also set those cells to protected to help prevent the cell from being overwritten.

eprbell replied:
I found a solution for this (let's consider line 4 as an example):
* Input tab: set B4 to `Enter Value...`
* Asset tab: set F4 to `=IF(IFNA(MATCH("Enter Value...", Input.B4), 0)=1, "Fill Input Sheet", Input.B4)`

mdavid217 replied:
Extending the formula's certainly an option too. Usually on ad hoc sheets I try to keep formulas as uncomplicated as possible, but for something generated by software, that's certainly possible.

We'd still need to use VLOOKUP since an asset with multiple holders (or multiple exchanges having balances on the Asset - Exchange tab) would break any relationship between row numbers on tabs. Essentially something like this:

`=IF( VLOOKUP(A4,$Input.A:B, 2, 0)="Enter value...", "Fill Input Sheet", VLOOKUP(A4,$Input.A:B, 2, 0))`

eprbell replied:
> 

I agree that since the output is generated we can use implementation techniques that otherwise it would be best to avoid: in a manually created spreadsheet I would also use VLOOKUP, but in generated files it's fine to use more complex formulas.

As for the `Asset - Exchange` tab, it's just a many to one relationship between (exchange, holder) -> asset. You can use VLOOKUP if you prefer, but I would suggest using a Python dictionary to keep track of asset -> row_number in the Input tab (so that you know which line an asset is generated at in the Input tab) and then use the dictionary to create the formulas of both the `Asset` and the `Asset - Exchange` tabs, without having to resort to VLOOKUP.

mdavid217 commented:
> One more thought: I can see this plugin evolving into something more than open positions. For example, it could offer advice on tax loss harvesting or other investment management issues. Should we give it a more open-ended name, like investment_report?

You're definitely correct about tax loss harvesting. All the information is there for someone to take advantage of already - pre-filtering that wouldn't be a big deal. It might be a lot trickier if (or when?) the US Gov decides to apply the wash sale rule to cryptos.

As far as the name, I'd consider all of the reports to be investment reports: transactions, summaries, realized gains, interest, airdrops, all the rest. To me, an unrealized gains / open positions report is just another investment report to go along with the rest. It's not a big deal to me to rename it, but I do feel at least for it's current contents it is reasonably named.

eprbell commented:
> 

Ok, sounds good, let's leave the name as is.

mdavid217 commented:
Hmm.. Not sure what happened on the checks there. Runs fine locally.

$ git status
On branch openpositions
Your branch is up to date with 'origin/openpositions'.

nothing to commit, working tree clean
$ mypy src tests
Success: no issues found in 57 source files`

eprbell commented:
Looks great: one more iteration and we're done!

The static analysis error is probably due to a bug in the latest mypy: can you merge with the latest RP2 code? I downgraded mypy to the previous version.

Also can you address the template file comments:
* cell "of Total Cost Basis %" is in a different font size: best to leave it in the same font size as the rest. If the user wishes, they can adjust column size with: Edit / Select All + Format / Columns / Optimal Width.
* legend: Fill in Asset Prices in the AssetPrice Tab for calculations -> Fill in Asset Prices in the Input Tab for calculations.
* legend: Information calculated based on values entered in AssetPrice tab. -> Information calculated based on values entered in Input tab

Finally add a small section for the plugin in docs/output_files.md: just a few lines describing what it does and directing the user to fill the input tab.

mdavid217 commented:
Ahh yes, meant to look at that column. Good catch on those legend changes too. All changes incorporated. Feel free to tweak docs and replace images to your liking - used an example file for it.

eprbell commented:
I think these things are still missing:
* cell "of Total Cost Basis %" is in a different font size: best to leave it in the same font size as the rest. If the user wishes, they can adjust column size with: Edit / Select All + Format / Columns / Optimal Width
* the static analysis error is probably due to a bug in the latest mypy: can you merge with the latest RP2 code? I downgraded mypy to the previous version.

mdavid217 commented:
Oops. Looks like I just corrected the font size on the asset/exch tab. Asset tab now updated also. 

Github indicates everything is up to date and doesn't seem to reflect your change to workflows. Once it does I can merge it in.

eprbell commented:
Ah, looks like I had fixed it in DaLI, but not in RP2: can you update your code now and try again?

eprbell commented:
Looks good to me: thanks for writing a great plugin!

mdavid217 commented:
Awesome! My pleasure. Hope others find it as useful as I do. :)
